### PR TITLE
Caching clang-tidy results

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use https://github.com/matus-chochlik/ctcache for clang-tidy results caching

### CODE REVIEW

File: cmake/clang_tidy.cmake [hunk 1]
```diff
@@ -3,10 +3,17 @@ option (ENABLE_CLANG_TIDY "Use clang-tidy static analyzer" OFF)
 
 if (ENABLE_CLANG_TIDY)
 
-    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
+    if (CLANG_TIDY_CACHE_PATH)
+        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+
+        set (CLANG_TIDY_PATH "${CLANG_TIDY_CACHE_PATH};${_CLANG_TIDY_PATH}" CACHE STRING "a doc")
```
Comment (alexey-milovidov) [inline cmake/clang_tidy.cmake:10]: `"a doc"` ?

Comment (Felixoid) [reply cmake/clang_tidy.cmake:10]: A good thing to fix in the next commit after the first clang-tidy is done :+1:

Comment (alexey-milovidov) [reply cmake/clang_tidy.cmake:10]: `${CLANG_TIDY_CACHE_PATH};${_CLANG_TIDY_PATH}` how does it work? Why do we concatenate it with `;`?

I'm not an expert in CMake... will it unfold to whitespace later?

File: docker/packager/binary/Dockerfile [hunk 2]
```diff
@@ -91,6 +91,9 @@ ENV PATH="$PATH:/usr/local/go/bin"
 ENV GOPATH=/workdir/go
 ENV GOCACHE=/workdir/
 
+RUN curl https://raw.githubusercontent.com/matus-chochlik/ctcache/1194532/clang-tidy-cache -Lo /usr/bin/clang-tidy-cache \
```
Comment (alexey-milovidov) [inline docker/packager/binary/Dockerfile:94]: `1194532` is it a commit sha prefix?
Looks unsafe.

Replace to the whole 160-bit commit sha.

Comment (alexey-milovidov) [reply docker/packager/binary/Dockerfile:94]: How do we check that we are on x86_64 here?

File: docker/packager/packager [hunk 3]
```diff
@@ -258,6 +258,7 @@ def parse_env_variables(
         if clang_tidy:
             # 15G is not enough for tidy build
             cache_maxsize = "25G"
+            result.append("CTCACHE_DIR=/ccache/clang-tidy")
```
Comment (alexey-milovidov) [inline docker/packager/packager:261]: What is the `CTCACHE_DIR` environment variable and where it is used?
How the `clang-tidy` program appears inside the `/ccache/` directory,
and why it is `clang-tidy`, not `clang-tidy-cache`?

Missing a comment.

Comment (Felixoid) [reply docker/packager/packager:261]: it's the directory for the cache itself. `CTCACHE_DIR` has the same purpose as the `CCACHE_DIR` above.

It's there to have the clang-tidy cache embedded into our standard `CCACHE_DIR`

File: cmake/clang_tidy.cmake (hunk 1 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:10]: It's a cmake black magic https://cmake.org/cmake/help/latest/prop_tgt/LANG_CLANG_TIDY.html#prop_tgt:%3CLANG%3E_CLANG_TIDY

the CLANG_TIDY_PATH is passed to CMAKE_CXX_CLANG_TIDY, which follows CXX_CLANG_TIDY syntax

**Comment by Felixoid** [comment]
It works. If the PRs https://github.com/matus-chochlik/ctcache/pull/5 and https://github.com/matus-chochlik/ctcache/pull/7 won't be merged we can for it into our organization.

Generally, I feel it should work!

File: docker/packager/binary/Dockerfile [hunk 4]
```diff
@@ -91,6 +91,9 @@ ENV PATH="$PATH:/usr/local/go/bin"
 ENV GOPATH=/workdir/go
 ENV GOCACHE=/workdir/
 
+RUN curl https://raw.githubusercontent.com/matus-chochlik/ctcache/1f6ce6b0581b00cffcd1f6642c7fd41cc81f2d21/clang-tidy-cache -Lo /usr/bin/clang-tidy-cache \
```
Comment (rschu1ze) [inline docker/packager/binary/Dockerfile]: I found no official releases for clang-tidy-cache: https://github.com/matus-chochlik/ctcache/releases

Just wondering how you found this binary?

File: docker/packager/packager (hunk 3 above)
Comment (rschu1ze) [reply docker/packager/packager:264]: It "feels" a bit weird to embed one cache directory into the other.  Sounds safer to have both side-by-side. Also, the name "clang-tidy-cache/" would be more fitting.

File: cmake/clang_tidy.cmake [hunk 5]
```diff
@@ -3,10 +3,20 @@ option (ENABLE_CLANG_TIDY "Use clang-tidy static analyzer" OFF)
 
 if (ENABLE_CLANG_TIDY)
 
-    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
+    if (CLANG_TIDY_CACHE_PATH)
+        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
```
Comment (rschu1ze) [inline cmake/clang_tidy.cmake:8]: No need to prefix with _, you could just do:
```cmake
find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
set (CLANG_TIDY_PATH "${CLANG_TIDY_CACHE_PATH};${CLANG_TIDY_PATH}" CACHE STRING "a doc")
```
(and pull out `find_program` so it is not replicated in both if branches.)

File: docker/packager/binary/Dockerfile [hunk 6]
```diff
@@ -91,6 +91,9 @@ ENV PATH="$PATH:/usr/local/go/bin"
 ENV GOPATH=/workdir/go
 ENV GOCACHE=/workdir/
 
+RUN curl https://raw.githubusercontent.com/matus-chochlik/ctcache/7fd516e91c17779cbc6fc18bd119313d9532dd90/clang-tidy-cache -Lo /usr/bin/clang-tidy-cache \
```
Comment (rschu1ze) [inline docker/packager/binary/Dockerfile:94]: I found no official releases for clang-tidy-cache: https://github.com/matus-chochlik/ctcache/releases

Just wondering how you found this binary?

File: docker/packager/packager [hunk 7]
```diff
@@ -258,6 +258,10 @@ def parse_env_variables(
         if clang_tidy:
             # 15G is not enough for tidy build
             cache_maxsize = "25G"
+
+            # `CTCACHE_DIR` has the same purpose as the `CCACHE_DIR` above.
+            # It's there to have the clang-tidy cache embedded into our standard `CCACHE_DIR`
+            result.append("CTCACHE_DIR=/ccache/clang-tidy")
```
Comment (rschu1ze) [inline docker/packager/packager:264]: It "feels" a bit weird to embed one cache directory into the other. Sounds safer to have both side-by-side. Also, the name "clang-tidy-cache/" would be more fitting.

File: cmake/clang_tidy.cmake [hunk 8]
```diff
@@ -3,10 +3,20 @@ option (ENABLE_CLANG_TIDY "Use clang-tidy static analyzer" OFF)
...
+
+        # Why do we use ';' here?
+        # It's a cmake black magic: https://cmake.org/cmake/help/latest/prop_tgt/LANG_CLANG_TIDY.html#prop_tgt:%3CLANG%3E_CLANG_TIDY
+        # The CLANG_TIDY_PATH is passed to CMAKE_CXX_CLANG_TIDY, which follows CXX_CLANG_TIDY syntax.
+        set (CLANG_TIDY_PATH "${CLANG_TIDY_CACHE_PATH};${_CLANG_TIDY_PATH}" CACHE STRING "A combined command to run clang-tidy with caching wrapper")
```
Comment (rschu1ze) [inline cmake/clang_tidy.cmake:13]: I don't think caching (CACHE STRING "...") is needed.

File: cmake/clang_tidy.cmake [hunk 9]
```diff
@@ -3,10 +3,20 @@ option (ENABLE_CLANG_TIDY "Use clang-tidy static analyzer" OFF)
 
 if (ENABLE_CLANG_TIDY)
 
-    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
+    if (CLANG_TIDY_CACHE_PATH)
+        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+
+        # Why do we use ';' here?
```
Comment (rschu1ze) [inline cmake/clang_tidy.cmake:10]: [CMake docs for set()](https://cmake.org/cmake/help/latest/command/set.html) say:

"Multiple arguments will be joined as a [semicolon-separated list](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-lists) to form the actual variable value to be set."

So you could just write `set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH})` and omit the comment about black magic.

File: docker/packager/binary/Dockerfile (hunk 4 above)
Comment (Felixoid) [reply docker/packager/binary/Dockerfile]: It's not the binary, it's a python script from a particular commit (current master)

Comment (Felixoid) [reply docker/packager/binary/Dockerfile]: Oh, no. This one is from https://github.com/Felixoid/ctcache/tree/all_merged

File: cmake/clang_tidy.cmake (hunk 5 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:8]: I am pretty sure I've tried it, but let's see if it works. Thanks!

File: docker/packager/binary/Dockerfile (hunk 6 above)
Comment (Felixoid) [reply docker/packager/binary/Dockerfile:94]: The current master, see https://github.com/matus-chochlik/ctcache/blob/7fd516e91c17779cbc6fc18bd119313d9532dd90/clang-tidy-cache

File: docker/packager/packager (hunk 7 above)
Comment (Felixoid) [reply docker/packager/packager:264]: No problem to rename, but the overhead to have another cache is too much. TOO much

File: cmake/clang_tidy.cmake (hunk 5 above)
Comment (rschu1ze) [reply cmake/clang_tidy.cmake:8]: I have tried it locally and the command line looked the same as before.

File: cmake/clang_tidy.cmake (hunk 9 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:10]: `${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH}` is not what I need.

File: docker/packager/binary/Dockerfile (hunk 6 above)
Comment (rschu1ze) [reply docker/packager/binary/Dockerfile:94]: Ah, didn't realize it's implemented in Python.

File: cmake/clang_tidy.cmake (hunk 9 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:10]:
```
    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    if (CLANG_TIDY_CACHE_PATH)
        set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${CLANG_TIDY_PATH})
    endif ()
```

```
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/felixoid/bin/clang-tidy-cache
CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy
```

It doesn't work

File: cmake/clang_tidy.cmake (hunk 5 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:8]: see https://github.com/ClickHouse/ClickHouse/pull/42913#discussion_r1012696241

File: cmake/clang_tidy.cmake (hunk 8 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:13]: Let's continue all questions regarding the block in one conversation, https://github.com/ClickHouse/ClickHouse/pull/42913#discussion_r1012696241

File: cmake/clang_tidy.cmake (hunk 9 above)
Comment (Felixoid) [reply cmake/clang_tidy.cmake:10]: And it neither

```
    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
    find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    if (CLANG_TIDY_CACHE_PATH)
        set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH})
    else ()
        set (CLANG_TIDY_PATH ${_CLANG_TIDY_PATH})
    endif ()
```

```
> cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=Debug -DSANITIZE= -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DENABLE_CLANG_TIDY=1 -DENABLE_TESTS=1 -DENABLE_EXAMPLES=1 -DENABLE_UTILS=1 .. | grep TIDY
By default, if no option is selected, on POSIX, libcurl transport adapter is used.
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/felixoid/bin/clang-tidy-cache
ENABLE_CLANG_TIDY:BOOL=1
_CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy
```

File: docker/packager/packager (hunk 7 above)
Comment (rschu1ze) [reply docker/packager/packager:264]: I am just afraid that ccache will mess up its own cached artefacts with the clang-tidy-cache ones.

**Review by rschu1ze — APPROVED** [review]

File: docker/packager/packager (hunk 7 above)
Comment (Felixoid) [reply docker/packager/packager]: Looks pretty good so far

https://github.com/ClickHouse/ClickHouse/actions/runs/3384444671/jobs/5621834614#step:7:575 +  
https://github.com/ClickHouse/ClickHouse/actions/runs/3384444671/jobs/5621834614#step:7:13233

File: cmake/clang_tidy.cmake (hunk 9 above)
Comment (rschu1ze) [reply cmake/clang_tidy.cmake:10]: To simplify this at least a little bit, I tried

```cmake
find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
if (CLANG_TIDY_CACHE_PATH)
    find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH} CACHE STRING "a doc")
else ()
    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
endif ()
```

this gives
```
By default, if no option is selected, on POSIX, libcurl transport adapter is used.
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/ubuntu/bin/clang-tidy-cache                  
CLANG_TIDY_PATH:STRING=/home/ubuntu/bin/clang-tidy-cache;/usr/bin/clang-tidy      
ENABLE_CLANG_TIDY:BOOL=1                                                          
_CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy                                     
```

**Comment by azat** [comment]
@Felixoid I haven't looked deeply enough, yet, but maybe you know the answer, why it still works too slow?

Right now it takes [~3 hours](https://s3.amazonaws.com/clickhouse-builds/46191/2f9ad860bd7e8ce59a996c762f855b33699f5473/clickhouse_special_build_check/report.html) with almost zero changes in sources (one .cpp module), while previously it takes [~4.5 hours](https://s3.amazonaws.com/clickhouse-builds/42190/bcc8d9f9d9570fdfab141d5830bef044e1676749/clickhouse_special_build_check/report.html), yes it works faster, but looks like it should be significantly faster.

**Comment by rschu1ze** [comment]
Runtimes of different clang-tidy checks differ a lot. I had some success a while ago with disabling certain expensive sub-checks (#43863). Unfortunately, I am not aware of logging or other tools in clang-tidy which would show how long individual checks run, i.e. "bisecting" the runtime is currently the only approach ...

**Comment by azat** [comment]
@rschu1ze Yeah, but still, in my example above, it takes **3 hours**, while there is **only one cpp module changed** (`S3Common.cpp`), so it should be more or less the same as running `clang-tidy` on this changed file

**Comment by rschu1ze** [comment]
Agree. Caching seems ineffective then.

**Comment by Felixoid** [comment]
clang-tidy-cache generates the hash for the cache out of the full source generated via `-E`, so if S3Common.cpp is used widely, it will change the hash for all dependent files. I think, it's the same logic as for ccache?

**Comment by azat** [comment]
>clang-tidy-cache generates the hash for the cache out of the full source generated via -E, so if S3Common.cpp is used widely

It cannot, it is not a header.

>it will change the hash for all dependent files. I think, it's the same logic as for ccache?

I wish :)

**Pull request merged**